### DISCUSSION TITLE
Time Out Issues

### DISCUSSION THREAD

**Discussion by ditchn** [belyaevvo/ClickHouseConnector#2] [Q&A] [answered]
Appreciate the effort developing this code.  I can connect to my source, select direct query mode, and apply.  but getting timeout errors when selecting fields to display.  I'm assuming that timing settings are a function of the config for the CH ODBC driver?  If not are there any other timing setting available?  

I've been consistently getting t/o errors after 30 seconds despite any changes made to the Timeout setting from within the ODBC data source admin panel.

...Any thoughts?

**Comment by ditchn** [comment]
to follow up on the error described above, feedback from the clickhouse teams after analyzing the logs the query is producing the error code:

Code: 24. DB::Exception: Cannot write to ostream at offset 230: While executing ODBCDriver2BlockOutputFormat. (CANNOT_WRITE_TO_OSTREAM).

This error means that a client who made a request is gone and Clickhouse unable to send the result (write the result into a network socket) because the client (receiver) has disappeared.

**Reply by belyaevvo** [reply, marked as answer]
Thank you for your explanations. I think you need to set the Timeout option in the connection string. I have made some modifications and published the latest release, so now you can download it and set the Timeout in seconds.

**Reply by ditchn** [reply]
Thanks for looking into the issue, I'll give that a shot, however, to be clear this is not a connection time-out.  

I'm convinced after additional testing/research, this is a **query timeout issue.**   I confirmed with the CH admin that after executing a direct query from PBI, CH continues to process the request well after the driver time-out.  This specific case CH is processing well over 5 Billion rows, taking roughly 5 minutes to execute, the driver consistently timesout at 30 seconds.  30 seconds is a default query/command timeout.  see MS doc for reference:  https://learn.microsoft.com/en-us/troubleshoot/sql/database-engine/connect/timeout-expired-error?source=recommendations

...it would seem any fix needs to address the call stack of a query command timeout.

again, appreciate your effort on this.  I'll run your revised code and send you an update.

**Reply by belyaevvo** [reply]
ClickHouse doesn't support separately command and connection timeout. It's seems that Timeout option control both connection and command timeouts. I've tested with play.clickhouse.com and ensure  that the timeout issues are resolved

**Comment by ditchn** [comment]
...issue solved, much appreciated.
